### PR TITLE
B0 Map correction added for when the unwrapping isn't complete

### PR DESCRIPTION
**Pull request by JSousa-UoL**
### Proposed changes

This Pull Request deals with the cases where the unwrapping of the phase images in B0 isn't complete. The bit of code written adds/subtracts `1/deltaTE` to the resulting B0Map. Please have a look at the changes and at the wording as well. This was tested with a dataset provided by @charlotteebuchanan

### CODE REVIEW

**Comment by JSousa-UoL** [comment]
We could close issue #42 after merging this PR

**Comment by alexdaniel654** [comment]
Hi Joao,

I've just taken a look at this, I have some concerns about using the central column of voxels as the baseline as they will always be in the spine. As the spine has a shorter T2* than the kidneys, it begins to have un-unwrappable voxels before the kidneys do when they do unwrap, they can end up multiple 2 pi away from the kidneys. This would result in a large B0 offset being applied to the kidneys to make the noise in the spine equal to zero. The example below is actually T2* mapping data but shows the wrapping in the spine before the kidneys/liver quite nicely.

image

**Comment by JSousa-UoL** [comment]
Hi Alex, I take your point and I think you are right.

The challenge here is to add or subtract `dynamic_range` (which is given by `1/delta`) to the resulting B0 Map. The question is how do you detect a large B0 Map offset and find out that there were multiple 2pi in the unwrapping? And if so, if it was -2pi or 2pi.

File: ukat/mapping/b0.py [hunk 1]
```diff
@@ -94,6 +94,17 @@ def __init__(self, pixel_array, echo_list, affine, mask=None,
             self.phase_difference = self.phase1 - self.phase0
             # B0 Map calculation
             self.b0_map = self.phase_difference / (2 * np.pi * self.delta_te)
+            # Check if the full cycle exists and if so, correct the output
+            # B0 Map based on the dynamic range
+            central_pixel_value = np.mean(self.b0_map[int(self.shape[0] / 2),
```
Comment (alexdaniel654) [inline ukat/mapping/b0.py:99]: To avoid the issues associated with taking a single column of values from the spine as the ground truth, zero offset, perhaps take the central x% of voxels where x is tuned to include a reasonable amount of kidney (and liver and spleen)? Something like below which uses the central 50% of voxels?

```python
prop = 0.5
central_pixel_value = np.mean(self.b0_map[int(self.b0_map.shape[0] / 2 - prop / 2 * self.b0_map.shape[0]):int(self.b0_map.shape[0] / 2 + prop / 2 * self.b0_map.shape[0]), int(self.b0_map.shape[1] / 2 - prop / 2 * self.b0_map.shape[1]):int(self.b0_map.shape[1] / 2 + prop / 2 * self.b0_map.shape[1]), :])
```
Also maybe this could be renamed to `mean_central_b0` or something similar, `central_pixel_value` isn't especially descriptive and it's a mean of values rather than a single value anyway.

File: ukat/mapping/b0.py [hunk 2]
```diff
@@ -94,6 +94,17 @@ def __init__(self, pixel_array, echo_list, affine, mask=None,
...
+                                                      int(self.shape[1] / 2),
+                                                      ...])
+            dynamic_range = 1 / self.delta_te
```
Comment (alexdaniel654) [inline ukat/mapping/b0.py:102]: Dynamic range normally means `voxel_array.max() - voxel_array.min()`. Perhaps rename to `b0_offset_step` or something?

File: ukat/mapping/b0.py [hunk 3]
```diff
@@ -94,6 +94,17 @@ def __init__(self, pixel_array, echo_list, affine, mask=None,
...
+            if np.abs(central_pixel_value) > dynamic_range / 2:
```
Comment (alexdaniel654) [inline ukat/mapping/b0.py:103]: Could you do something along the lines of 
```python
self.b0_map -= (central_pixel_value // dynamic_range) * dynamic_range
```
to replace 103 - 107. It's more succinct and can deal with cases where central_pixel_value is more than one multiple of dynamic_range away from zero.

**Review by alexdaniel654 — CHANGES_REQUESTED** [review]
Hopefully some useful inline comments. It might be a bit clearer what's going on if this correction is done to `phase_difference` as some of the variables make a bit more physical sense there. `central_pixel_value` would become `central_phase_difference` and `dynamic_range` would just be 2 * pi. It certainly all sits nicer in my mind when I'm thinking of it as phases rather than Hz, but that could be just me.

**Comment by JSousa-UoL** [comment]
Thank you so much for your feedback and your help Alex! It's much clearer now to me and the suggestions you made are implemented and they look a lot cleaner.

**Review by alexdaniel654 — CHANGES_REQUESTED** [review]
Looks good, is it possible to add some dedicated tests where `mean_central_b0 // b0_offset_step` equals zero and not zero e.g. `test_b0_offset_removal`. Currently nothing actually checks this PR does what's expected, it just checks it runs with the test data, none of which has a B0 offset (otherwise tests would have failed when this feature was added).

**Comment by JSousa-UoL** [comment]
I was wondering about that, because I tested locally but that means nothing really. I guess I'll have to add an example that has the B0 offset in the Zenodo

**Review by alexdaniel654 — APPROVED** [review]
That looks good to me. 👍
